java:app & java:module namespaces temporarily disappear when SAM invoked

When trying to obtain something from the JNDI namespace <tt>java:app</tt> or <tt>java:module</tt> during invocation of a SAM's <tt>validateRequest</tt> method an exception is thrown.

However, right before the SAM is invoked, namely in a <tt>ServletRequestListener</tt> as well as right after the SAM is invoked, namely in a Filter or Servlet this operation succeeds.

The exception thrown is the following:

```
javax.naming.NamingException: Lookup failed for 'java:module/EJBBean' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming} [Root exception is javax.naming.NamingException: Invocation exception: Got null ComponentInvocation ]
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:491)
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:438)
	at javax.naming.InitialContext.lookup(InitialContext.java:417)
	at javax.naming.InitialContext.lookup(InitialContext.java:417)
	at test.TestServerAuthModule.validateRequest(TestServerAuthModule.java:40)
	at javax.security.authenticationmechanism.DefaultServerAuthContext.validateRequest(DefaultServerAuthContext.java:36)
	at com.sun.web.security.RealmAdapter.validate(RealmAdapter.java:1654)
	at com.sun.web.security.RealmAdapter.invokeAuthenticateDelegate(RealmAdapter.java:1521)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:606)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:702)
	at org.apache.catalina.core.StandardPipeline.doChainInvoke(StandardPipeline.java:678)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:97)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:416)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:283)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:206)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:180)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:283)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:200)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:132)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:111)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:536)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:591)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:571)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.naming.NamingException: Invocation exception: Got null ComponentInvocation 
	at com.sun.enterprise.naming.impl.GlassfishNamingManagerImpl.getComponentId(GlassfishNamingManagerImpl.java:842)
	at com.sun.enterprise.naming.impl.GlassfishNamingManagerImpl.lookup(GlassfishNamingManagerImpl.java:714)
	at com.sun.enterprise.naming.impl.JavaURLContext.lookup(JavaURLContext.java:167)
	at com.sun.enterprise.naming.impl.SerialContext.lookup(SerialContext.java:471)
```

The lookup code is as follows:

```
try {
	// java:global/glassfish-sam-ee-namespaces/EJBBean does work 	EJBBean ejbBean = (EJBBean) new InitialContext().lookup("java:module/EJBBean");
	System.out.println("SAM: " + ejbBean.hello());
} catch (NamingException e) {
	System.out.println("SAM: Exception");
	e.printStackTrace();
}
```

A full reproducer is provided as Maven project at: [https://github.com/arjantijms/glassfish-sam-ee-namespaces](https://github.com/arjantijms/glassfish-sam-ee-namespaces)

Building and deploying this project to a stock GlassFish, then requesting [http://localhost:8080/glassfish-sam-ee-namespaces/servlet](http://localhost:8080/glassfish-sam-ee-namespaces/servlet) prints the following in the log:

```
RequestListener: Hello from EJB
SAM: Exception
javax.naming.NamingException: Lookup failed for 'java:module/EJBBean' in SerialContext[myEnv={java.naming.factory.initial=com.sun.enterprise.naming.impl.SerialInitContextFactory, java.naming.factory.state=com.sun.corba.ee.impl.presentation.rmi.JNDIStateFactoryImpl, java.naming.factory.url.pkgs=com.sun.enterprise.naming} (rest of exception as above)
Servlet: Hello from EJB
```

Expected is:

```
RequestListener: Hello from EJB
SAM: Hello from EJB
Servlet: Hello from EJB
```

Deploying the same application to JBoss WildFly 9.0.1 or 10.0rc3 (with a modified standalone.xml to activate jaspic) indeed prints the above.
#### Affected Versions
[4.1.1, 5.0]

--- a/main/appserver/security/webintegration/src/main/java/com/sun/web/security/RealmAdapter.java
+++ b/main/appserver/security/webintegration/src/main/java/com/sun/web/security/RealmAdapter.java
@@ -87,7 +87,7 @@ import org.apache.catalina.deploy.LoginConfig;
 import org.apache.catalina.deploy.SecurityConstraint;
 import org.apache.catalina.realm.Constants;
 import org.apache.catalina.realm.RealmBase;
-
+import org.apache.catalina.ContainerEvent;
 import org.glassfish.api.invocation.ComponentInvocation;
 import org.glassfish.internal.api.ServerContext;
 //import com.sun.enterprise.Switch;
@@ -473,7 +473,12 @@ public class RealmAdapter extends RealmBase implements RealmInitializer, PostCon
                     if (subject.isReadOnly()) {
                         _logger.log(Level.WARNING, "Read-only subject found during logout processing");
                     }
-                    sAC.cleanSubject(messageInfo, subject);
+                    try {
+                        req.getContext().fireContainerEvent(ContainerEvent.BEFORE_LOGOUT, null);
+                        sAC.cleanSubject(messageInfo, subject);
+                    }finally {
+                        req.getContext().fireContainerEvent(ContainerEvent.AFTER_LOGOUT, null);
+                    }
                 }
             } catch (AuthException ex) {
                 throw new RuntimeException(ex);
@@ -1518,7 +1523,12 @@ public class RealmAdapter extends RealmBase implements RealmInitializer, PostCon
         
         if (serverAuthConfig != null) {
             //JSR 196 is enabled for this application
-            result = validate(request, response, config, authenticator, calledFromAuthenticate);
+            try{
+                context.fireContainerEvent(ContainerEvent.BEFORE_AUTHENTICATION, null);
+                result = validate(request, response, config, authenticator, calledFromAuthenticate);
+            }finally {
+                context.fireContainerEvent(ContainerEvent.AFTER_AUTHENTICATION, null);
+            }
         } else {
             //jsr196 is not enabled.  Use the current authenticator.
             result = ((AuthenticatorBase) authenticator).authenticate(
@@ -1551,10 +1561,15 @@ public class RealmAdapter extends RealmBase implements RealmInitializer, PostCon
                     //JSR 196 is enabled for this application
                     sAC = (ServerAuthContext) messageInfo.getMap().get(SERVER_AUTH_CONTEXT);
                     if (sAC != null) {
-                        AuthStatus authStatus =
-                                sAC.secureResponse(messageInfo,
-                                null); //null serviceSubject
-                        result = AuthStatus.SUCCESS.equals(authStatus);
+                        try {
+                            context.fireContainerEvent(ContainerEvent.BEFORE_POST_AUTHENTICATION, null);
+                            AuthStatus authStatus =
+                                    sAC.secureResponse(messageInfo,
+                                            null); //null serviceSubject
+                            result = AuthStatus.SUCCESS.equals(authStatus);
+                        }finally {
+                            context.fireContainerEvent(ContainerEvent.AFTER_POST_AUTHENTICATION, null);
+                        }
                     }
                 }
             }

--- a/main/appserver/web/web-core/src/main/java/org/apache/catalina/ContainerEvent.java
+++ b/main/appserver/web/web-core/src/main/java/org/apache/catalina/ContainerEvent.java
@@ -211,6 +211,19 @@ public final class ContainerEvent extends EventObject {
     public static final String AFTER_WRITE_LISTENER_ON_ERROR
         = "afterWriteListenerOnError";
 
+    public static final String BEFORE_AUTHENTICATION
+        = "beforeAuthentication";
+    public static final String AFTER_AUTHENTICATION
+        = "afterAuthentication";
+    public static final String BEFORE_POST_AUTHENTICATION
+        = "beforePostAuthentication";
+    public static final String AFTER_POST_AUTHENTICATION
+        = "afterPostAuthentication";
+    public static final String BEFORE_LOGOUT
+        = "beforePostAuthentication";
+    public static final String AFTER_LOGOUT
+        = "afterPostAuthentication";
+
     /**
      * The Container on which this event occurred.
      */

--- a/main/appserver/web/web-glue/src/main/java/com/sun/web/server/WebContainerListener.java
+++ b/main/appserver/web/web-glue/src/main/java/com/sun/web/server/WebContainerListener.java
@@ -100,6 +100,9 @@ public final class WebContainerListener
         beforeEvents.add(ContainerEvent.BEFORE_READ_LISTENER_ON_ERROR);
         beforeEvents.add(ContainerEvent.BEFORE_WRITE_LISTENER_ON_WRITE_POSSIBLE);
         beforeEvents.add(ContainerEvent.BEFORE_WRITE_LISTENER_ON_ERROR);
+        beforeEvents.add(ContainerEvent.BEFORE_AUTHENTICATION);
+        beforeEvents.add(ContainerEvent.BEFORE_POST_AUTHENTICATION);
+        beforeEvents.add(ContainerEvent.BEFORE_LOGOUT);
 
         // postInvoke events
         afterEvents.add(ContainerEvent.AFTER_CONTEXT_INITIALIZED);
@@ -125,6 +128,9 @@ public final class WebContainerListener
         afterEvents.add(ContainerEvent.AFTER_READ_LISTENER_ON_ERROR);
         afterEvents.add(ContainerEvent.AFTER_WRITE_LISTENER_ON_WRITE_POSSIBLE);
         afterEvents.add(ContainerEvent.AFTER_WRITE_LISTENER_ON_ERROR);
+        afterEvents.add(ContainerEvent.AFTER_AUTHENTICATION);
+        afterEvents.add(ContainerEvent.AFTER_POST_AUTHENTICATION);
+        afterEvents.add(ContainerEvent.AFTER_LOGOUT);
 
     }
 

